### DISCUSSION TITLE
How do you check the Backup Eligibility Flag, or Backup State flag?

### DISCUSSION THREAD

**Discussion by anthonyryan1** [web-auth/webauthn-framework#422] [Q&A]
With Google and Apple now synchronizing private keys between devices. There is now a Backup Eligibility and Backup State flag in the specification.

https://w3c.github.io/webauthn/#backup-eligibility
https://w3c.github.io/webauthn/#backup-state

How would one go about checking for that flag with this library?

It could be useful from a UI perspective to know a private key will be available from many devices. (e.g. Label it as an "Google Passkey" instead of a "Pixel 5a Passkey").

**Comment by Spomky** [comment]
Hi,

At the moment, there is no convenient method to get these two values, but it should still be possible.
Both `AuthenticatorAssertionResponseValidator::check` and `AuthenticatorAttestationResponseValidator::check` methods emit an event in case of Assertion/Attestation validation success. This event contains the authenticator response and allows the access to the authenticator data

* Attestation ceremony: the authenticator data is in AuthenticatorAttestationResponse->[getAttestationObject()](https://github.com/web-auth/webauthn-framework/blob/4.6.x/src/webauthn/src/AuthenticatorAttestationResponse.php#L21)->[getAuthData()](https://github.com/web-auth/webauthn-framework/blob/4.6.x/src/webauthn/src/AttestationStatement/AttestationObject.php#L36)
* Assertion ceremony: the authenticator data is in AuthenticatorAssertionResponse->[getAuthenticatorData()](https://github.com/web-auth/webauthn-framework/blob/4.6.x/src/webauthn/src/AuthenticatorAssertionResponse.php#L23)

With the authenticator data object, you can call ->[getReservedForFutureUse2()](https://github.com/web-auth/webauthn-framework/blob/4.6.x/src/webauthn/src/AuthenticatorData.php#L72), which returns the flag bits [3, 4 and 5](https://github.com/web-auth/webauthn-framework/blob/4.6.x/src/webauthn/src/AuthenticatorData.php#L21).
The BE and BS flags are at bits 3 and 4 respectively.

I will add methods for simplifying this.

**Comment by Spomky** [comment]
FYI, it is now implemented in the next minor version 4.7.x and it works fine.
I prepared the demo application available at https://webauthn.spomky-labs.com/ and it is able to detect the credential type 🎉
![Capture d'écran 2023-06-21 221019](https://github.com/web-auth/webauthn-framework/assets/1091072/a9172468-0713-4bf7-b247-d9edde3daa78)
